Added support for an httpd authentication configuration map

abellotti opened the pull request:
Added support for an authentication config file directory is mounted from a configmap.

- Authentication Configuration Map mounted in:
    /etc/httpd/auth-conf.d/

- Config file driving the initialization is:
    /etc/httpd/auth-conf.d/auth-configuration.conf

- New script to overlay the external authentication files is:
    /usr/bin/initialize-httpd-auth.sh and invoked from the entrypoint.

Needs to be merged with: https://github.com/ManageIQ/manageiq-pods/pull/201

abellotti committed:
Added support for an authentication config file directory
that is mounted from a configmap.

- Authentication Configuration Map mounted in:
    /etc/httpd/auth-conf.d/

- Config file driving the initialization is:
    /etc/httpd/auth-conf.d/auth-configuration.conf

- New script to overlay the external authentication files is:
    /usr/bin/initialize-httpd-auth.sh and invoked from the entrypoint.

abellotti commented:
This capability allows for authentication to be configured/generated outside this Pod.  config maps can be auto-generated by scripts from other pods, updating the httpd-auth-configs one and then the httpd pod just recycled to honor the new authentication.

This also allows customers to easily update/customize their sssd.conf/any other conf files, then recycle the httpd pod.

Default behavior is the internal mode, as shown in the following log screen:
![screen shot 2017-08-17 at 5 20 23 pm](https://user-images.githubusercontent.com/5797328/29434721-3a222b26-8372-11e7-8faf-f342c6febeff.png)


With a run using sample data:
![screen shot 2017-08-17 at 5 17 56 pm](https://user-images.githubusercontent.com/5797328/29434741-4b85c724-8372-11e7-99b2-0c9e127baf1b.png)

Provision for simple and binary files, and arbitrary file paths.

carbonin commented on docker-assets/initialize-httpd-auth.sh:
```diff
@@ -0,0 +1,73 @@
+
+HTTPD_AUTH_CONFIG_DIR=/etc/httpd/auth-conf.d
+AUTH_CONFIG_FILE="${HTTPD_AUTH_CONFIG_DIR}/auth-configuration.conf"
+SSSD_CONF_FILE="/etc/sssd/sssd.conf"
```
If we are expecting specific file paths for particular authentication types, should we validate that the required files are present based on the type requested or is that too much to keep track of?

abellotti replied:
Not a question of keeping track of, but many authentication type do configure sssd, so it's file specific, not type specific. Other types in future could also require SSSD.

fbladilo replied:
@abellotti @carbonin Could we put the auth initialization scripts on their own systemd service unit like we use to do in fine : 

https://github.com/ManageIQ/manageiq-pods/blob/fine/images/miq-app/docker-assets/appliance-initialize.service

This will allow to free ENTRYPOINT and support OCI systemd hooks properly.

abellotti replied:
systemd service for this is fine, but what about other services that require an environment variable for startup (see upcoming auth-api service PR). https://github.com/abellotti/container-httpd/blob/d1f1141cbc384f22536c26e7e1b511886baffb98/docker-assets/entrypoint need a way to save the env before starting systemd.

fbladilo replied:
@abellotti To dump all environment to a file before systemd gets in, we can use a kube postStart lifecycle hook on the httpd DC, it takes a command or a script/entrypoint. 

We already use some lifecycle hooks across the deployment, it makes sense for something simple like this. Consider the following gist : 

https://gist.github.com/fbladilo/83d53f104985e5d48debae58ab6b5e25

carbonin commented on docker-assets/initialize-httpd-auth.sh:
```diff
@@ -0,0 +1,73 @@
+function auth_type {
+  if [ -f ${AUTH_CONFIG_FILE} ]
+  then
+    CONFIG_AUTH_TYPE=$(grep '^type[[:space:]]*=.*$' ${AUTH_CONFIG_FILE} | cut -f2 -d= | trim_whitespace)
```
This type of stuff really makes me want to have specs for something like this ...

Not sure if this would really work, but if we're going to have crystal on this pod anyway could we write this code up in a small crystal program so it would be easier to test? (This could also be a future thing, seeing as how we need to get auth in for the next release).

abellotti replied:
👍 on tests at some point.  Let's revisit once (if) crystal is in.

carbonin reviewed (approved):
A few concerns, but nothing major. Looks good to me.

abellotti committed:
- Updated the auth-config scripts to run as a service and relying
  on the /usr/bin/entrypoint which must be run by a postStart
  lifecycle hook.

- Leverage the /etc/container-environment file for the
  initialize-httpd-auth and httpd services.

carbonin commented on Dockerfile:
```diff
@@ -61,11 +61,23 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
+## Create the mount point for the authentication configuration files
+RUN mkdir /etc/httpd/auth-conf.d
+
+COPY docker-assets/entrypoint               /usr/bin
```
Can we call this something more descriptive as it's no longer the entrypoint of the continer?

Fryguy replied:
I was thinking the same thing as I read it, and just thought I was wrong.

carbonin commented on docker-assets/httpd-environment.conf:
```diff
@@ -0,0 +1,3 @@
+[Service]
+ConditionPathExists=/etc/container-environment
```
What environment variables need to be provided to the httpd service?

The only one I see in the other PR is `HTTPD_AUTH_TYPE` is that used by httpd itself?

carbonin replied:
Spoke with @abellotti, this is for future use as some of the config files will require environment variable interpolation.

abellotti committed:
Renamed httpd pod postStart script from entrypoint to save-container-environment

carbonin reviewed (approved):
Looks good to me.

I think we should coordinate merging this and https://github.com/ManageIQ/manageiq-pods/pull/201 as things will break if they are merged separately. Additionally because docker hub takes so long to build images, I'll probably build this image locally and push to ManageIQ/httpd after this is merged.

bdunne commented on docker-assets/initialize-httpd-auth.sh:
```diff
@@ -0,0 +1,65 @@
+  if [ $SSSD_CONF == "yes" ]
+  then
+    echo "SSSD File ${SSSD_CONF_FILE} copied, Enabling SSSD ..."
+    systemctl enable sssd
```
Does it also need to `start sssd`?

abellotti replied:
I didn't have to. This is run early, "Before=network-pre.target", systemd starts up dbus, sssd, httpd afterward.
